### PR TITLE
cabal: Submission of Haskell Cabal: Common Architecture for Building App…

### PR DESCRIPTION
**Pull request by essandess**
Submission cabal: Haskell Cabal: Common Architecture for Building Applications

* Bootsrapped build of cabal

###### Type(s)

- [x] submission
- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.14.5 18F132
Xcode 10.2.1 10E1001 

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by essandess** [comment]
This is weird: the new `ghc` isn’t being built as a dependency because `autoreconf` isn’t installed. See https://paste.z0k.xyz/1c452840a3e6.txt:
```
./boot  
sh: autoreconf: command not found
```

I presume the fix is easy. Would someone please post the necessary Portfile invocation to get this?

**Comment by essandess** [comment]
@cjones051073 Would you please take a look at this, and perhaps let me know if I missed something in the `ghc` Portfile that breaks when it’s built as a dependency?

**Comment by amake** [comment]
The Cabal 2.4.0.1 library appears to be installed with ghc 8.6.5: see the output of `port contents ghc`. Additionally cabal-install is supposed to be part of the haskell-platform port, but this hasn't been updated  along with ghc.

I suspect that this separate port isn't necessary.

**Comment by amake** [comment]
See also #5050

**Comment by essandess** [comment]
ghc has some cabal libraries, but not the `cabal` binary. This port bootstraps the binary. That’s why a separate port and build are necessary.

Also, one can install cabal with `cabal-install`, but that just downloads a binary without building or bootstrapping. It’s preferable to build bootstrapped binaries so long as it’s not too difficult to fragile. 

> The Cabal 2.4.0.1 library appears to be installed with ghc 8.6.5: see the output of `port contents ghc`. Additionally cabal-install is supposed to be part of the haskell-platform port, but this hasn't been updated along with ghc.
> 
> I suspect that this separate port isn't necessary.

**Comment by essandess** [comment]
For some reason the buildbot isn’t finding the `autoreconf` binary when ‘port:ghc` is included as a dependency. It builds fine otherwise. See https://paste.z0k.xyz/7c9da66b4c54. This happens even if `port:autoconf` Is listed explicitly in `depends_build`:
```
DEBUG: system -W /opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_lang_ghc/ghc/work/src/ghc-8.6.5:  export  PATH=/opt/local/bin:/opt/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_lang_ghc/ghc/work/bootstrap/bin ;  ./boot  
sh: autoreconf: command not found
```

Perhaps some subtle issue with `PATH and the python3 script `./boot`? I don’t see the issue myself now.

**Comment by essandess** [comment]
@cjones051073 @mf2k @neverpanic I believe this is ready, subject to any review and comments, but two of the azure buildbots appear to have hung. What’s the best approach to move forward? Just kick it again with any empty commit?

**Comment by cjones051073** [comment]
The azure builds can be restarted. Just cluck on the details link.

**Comment by essandess** [comment]
@cjones051073 Thanks for restarting these. I didn’t know one could do that.

This PR is ready to merge.

**Comment by essandess** [comment]
@cjones051073 @mf2k Ready to merge, incorporates comments from PR https://github.com/macports/macports-ports/pull/5272.

File: lang/cabal/Portfile [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+# Bootstrapped cabal is the default build
+variant prebuilt \
+	    description {Do not bootstrap cabal; install the pre-built binary.} {}
+
+# use_xz              yes
+
+master_sites        https://downloads.haskell.org/~${name}/${name}-install-latest/
+distfiles           ${name}-install-${version}-x86_64-apple-darwin17.7.0.tar.xz
+distname            [join [lrange [split [lindex ${distfiles} end] .] 0 end-2] .]
```
Comment (mf2k) [inline lang/cabal/Portfile]: I believe that this is there wrong way. `distfiles` is automatically generated as `${distname}${extract.suffix}` and should never be manually set, especially using `distfiles`. Set the `distname` and `extract.suffix` directly instead. 

```
distname           ${name}-install-${version}-x86_64-apple-darwin17.7.0 
extract.suffix    tar.xz

```

Comment (essandess) [reply lang/cabal/Portfile]: Tanks! Fixed.

**Pull request merged**